### PR TITLE
[WIP] Simple head pat

### PR DESCRIPTION
**Pull request by trefis**

### CODE REVIEW

**Comment by gasche** [comment]
I don't like the name Simple_head_pat -- I preferred "discriminants" as we had before. I think that everything we call "pattern" should really be a pattern. "Simple patterns" or "head patterns" may have restrictions on the allowed constructors (no or, no alias, etc.) or their arguments, or both, but they should really be pattern. Here we expose a `desc` interface that is really not a pattern but the "spine" of a pattern. We ensure that the `t` are isomorphic to "omega patterns" (we should use this name!), but omega patterns are not the things we are manipulating here.

To make an analogy, those discriminants play the same role as "signs" for numbers: we group patterns according to those discriminants, just like we could partition a set of numbers into the positive and the negative numbers. It is possible to implement an operation `to_number : sign -> number`, which returns `+1` or `-1`, but signs are not numbers.

For that reason I think that any name that *ends* with "pattern" should be rejected. Here are some ideas that could be accepted:

- discriminant
- classifier
- spine/skeleton (not sure they give the right intuition)
- pattern head (and not "head pattern")
- pattern group
- pattern class

Relatedly, I think that having `of_pattern : pattern -> t` and `to_pattern : t -> pattern` is misleading, because it gives the impression that it is possible to convert back-and-forth to patterns (and then one intuitively thinks: without losing information). I would rename the latter into `to_omega_pattern : t -> pattern`. We could also consider having a different API, for example:

```ocaml
decompose/split : pattern -> t * pattern list
compose/plug/join : t * pattern list -> pattern
```

File: typing/parmatch.ml [hunk 1]
```diff
@@ -44,6 +44,132 @@ let omega_list l = List.map (fun _ -> omega) l
...
+
+  (** reconstructs a pattern, putting wildcards as sub-patterns. *)
+  val to_pattern : t -> pattern
+
+  val make_record
+    :  loc:Location.t
+    -> typ:Types.type_expr
+    -> env:Env.t
+    -> label_description list
+    -> t
```
Comment (gasche) [inline typing/parmatch.ml:81]: I'm not fond of this function. At first I thought that `_record` here referred to the internal implementation of `t` as a record. Couldn't we just have the more general

```ocaml
val make : loc -> typ -> env -> desc -> t
```

?

File: typing/parmatch.ml [hunk 2]
```diff
@@ -771,56 +866,51 @@ let close_variant env row =
...
+  | Variant _ ->
+      let fields =
+        List.map
+          (fun (d, _) ->
+            match Simple_head_pat.desc d with
+            | Variant { tag } -> tag
+            | _ -> assert false)
+          env
+      in
+      let row = Simple_head_pat.row_of_discr discr in
```
Comment (gasche) [inline typing/parmatch.ml]: Why not get the `row` field of the variant description here?

File: typing/parmatch.ml [hunk 3]
```diff
@@ -44,6 +44,132 @@ let omega_list l = List.map (fun _ -> omega) l
...
+
+  let desc { desc } = desc
+  let env { env } = env
+  let loc { loc } = loc
+  let typ { typ } = typ
+
+  let row_of_discr d =
+    match Ctype.expand_head d.env d.typ with
+    | {desc = Tvariant row} -> Btype.row_repr row
+    | _ -> assert false
```
Comment (gasche) [inline typing/parmatch.ml]: This function is not very nice. Why don't we just let the user match the `desc`, which should be a `Variant`, and access the `row` value there? (See comment at the use-site below.)

**Comment by trefis** [comment]
This is all seems very reasonable. I think I like "pattern head" more than "discriminant".
I will make that change if that's alright with you.

---

The idea of having a `decompose` function instead (or alongside) `of_pattern` is an interesting idea. In particular because I could see myself putting these pairs of `Pattern_head.t * pattern list` on the first column of simplified matrices, in constrast to the actual patterns I'm currently keeping there.

The "compose" operation is currently being called `combine` in Luc's papers, and perhaps we want to reuse that?

I don't have a strong opinion on which names we should use for these operations, but I'm keen on experimenting with that API.

**Review by gasche — COMMENTED** [review]
(bunch of minor comments)

**Comment by gasche** [comment]
I don't attach much value to reusing exactly the names in the paper, but if that can help make a choice about equally-okay proponents I'm all for it. Would it be `combine / decombine`?  (`uncombine` also exists, both are really rare, but `decombine` reads better to me and it had such a [huge](https://books.google.com/ngrams/graph?content=uncombine%2Cdecombine&year_start=1800&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t1%3B%2Cuncombine%3B%2Cc0%3B.t1%3B%2Cdecombine%3B%2Cc0#t1%3B%2Cuncombine%3B%2Cc0%3B.t1%3B%2Cdecombine%3B%2Cc0) popularity spike in 1830 it would give us a definite retro feel.)

File: typing/parmatch.ml (hunk 3 above)
Comment (trefis) [reply typing/parmatch.ml]: There is a two-parts answer to that question:
1. This function already existed: I merely reproduced the behavior of the existing code. So the question should really be directed at the (author of the) existing code. Of course we could take this PR as an opportunity to clean things up, however:
2.  a) I am not sure that the row you will find in the type is the same row that you will find on in `Tpat_variant`. Naively one would hope that they are the same (and there are good reasons to think they are), but I'm just not sure.
  b) In an initial version of the patch, I had made this simplifications, and then either the compiler failed to build itself, or the testsuite was failing (I can't quite remember). It's possible that the simplification was not to blame, that I had made another silly mistake. But the issue just isn't there without the "simplification".

**Comment by trefis** [comment]
I'm not sure we should strive for perfect symmetry in the naming. `decompose / combine` feels like the more natural verbs here.

File: typing/parmatch.ml (hunk 1 above)
Comment (trefis) [reply typing/parmatch.ml:81]: > At first I thought that _record here referred to the internal implementation of t as a record

That would have been a pretty bad name.

---

We could have the more general function sure, `make_record` does feel weird.

**Comment by gasche** [comment]
Yes, perfect symmetry! compose/decompose or combine/decombine.

File: typing/parmatch.ml (hunk 3 above)
Comment (gasche) [reply typing/parmatch.ml]: Uh, this sounds scary. I understand why you don't want to touch it anymore. Could you maybe try to check once again that using the `Variant`'s row at the use-site causes an issue, and then add a comment there to this effect?

**Comment by trefis** [comment]
> Yes, perfect symmetry! compose/decompose or combine/decombine.

The british person sitting next of me says "compose doesn't work". I didn't ask for his opinion on "decombine" which very clearly doesn't work.
That person suggests "deconstruct / construct" which I can get behind.

File: typing/parmatch.ml (hunk 3 above)
Comment (trefis) [reply typing/parmatch.ml]:
```
$ make world.opt

...

File ".../utils/profile.ml", line 239, characters 15-424:
239 | ...............(function
240 |         | `Time ->
241 |           make p.duration ~f:time_display
242 |         | `Alloc ->
243 |           make p.allocated_words ~f:memory_word_display
...
246 |         | `Abs_top_heap ->
247 |           make (float_of_int top_heap_words)
248 |            ~f:(memory_word_display ~previous:(float_of_int prev_top_heap_words))
249 |       ).........
Error (warning 8): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
`AnyOtherTag
```

Clearly, that isn't right.

The patch is here: https://gist.github.com/trefis/d6024cb33c5a84681e449e3741d3b641

**Comment by trefis** [comment]
I pushed a few more commits implementing some of the changes discussed here. The last one (introducing a `deconstruct` function) was clearly a fruitful idea: it allowed removing 6 `assert false`.
I will now rebase and integrate your changes regarding variant rows.

At which point I think there will be more than enough for a PR.

I haven't yet tried the idea of `Pattern_head.construct` instead of the current `set_args ∘ Pattern_head.to_omega_pattern`, but I think this could wait for a latter date. (In particular because there are interactions with `Matching` in this area)

**Comment by trefis** [comment]
I cherry-picked your changes. I think if I submit this upstream, I will squash all the commits into one beforehand.

**Comment by gasche** [comment]
Yes, I think this is good to send as a PR upstream. Maybe cleanup the history a bit to hide the temporary names?

**Comment by trefis** [comment]
Done.

**Pull request closed without merging**